packages/valory/contracts/gnosis_safe/contract.py

```diff
@@ -774,3 +775,192 @@ def get_safe_txs(
+        safe_contract = cls.get_instance(ledger_api, contract_address)
+        # Note that owners in the safe are stored as a linked list, we need to know the parent (prev_owner) of an owner
+        # when removing. https://github.com/safe-global/safe-contracts/blob/v1.3.0/contracts/base/OwnerManager.sol#L15
+        owners = [
+            ledger_api.api.toChecksumAddress(owner)
+            for owner in safe_contract.functions.getOwners().call()
+        ]
+        owner = ledger_api.api.toChecksumAddress(owner)
+        prev_owner = cls._get_prev_owner(owners, owner)
+        data = safe_contract.encodeABI(
+            fn_name="removeOwner",
+            args=[
+                ledger_api.api.toChecksumAddress(prev_owner),
+                owner,
+                threshold,
+            ],
+        )
+        return dict(
+            data=data,
+        )
+
+    @classmethod
+    def get_swap_owner_data(
+        cls,
+        ledger_api: EthereumApi,
+        contract_address: str,
+        old_owner: str,
+        new_owner: str,
+    ) -> JSONLike:
+        """
+        Get a swapOwner() encoded tx.
+
+        This method acts as a wrapper for `swapOwner()`
+        https://github.com/safe-global/safe-contracts/blob/v1.3.0/contracts/base/OwnerManager.sol#L94
+
+        :param ledger_api: the ledger API object
+        :param contract_address: the contract address
+        :param old_owner: the owner to be replaced
+        :param new_owner: owner to replace old_owner
+        :return: the zero transfer events
+        """
+        ledger_api = cast(EthereumApi, ledger_api)
+        safe_contract = cls.get_instance(ledger_api, contract_address)
+        # Note that owners in the safe are stored as a linked list, we need to know the parent (prev_owner) of an owner
+        # when swapping. https://github.com/safe-global/safe-contracts/blob/v1.3.0/contracts/base/OwnerManager.sol#L15
+        owners = [
+            ledger_api.api.toChecksumAddress(owner)
+            for owner in safe_contract.functions.getOwners().call()
+        ]
+        old_owner = ledger_api.api.toChecksumAddress(old_owner)
+        prev_owner = cls._get_prev_owner(owners, old_owner)
+        data = safe_contract.encodeABI(
+            fn_name="swapOwner",
+            args=[
+                ledger_api.api.toChecksumAddress(prev_owner),
+                old_owner,
+                ledger_api.api.toChecksumAddress(new_owner),
+            ],
+        )
+        return dict(
+            data=data,
+        )
+
+    @classmethod
+    def _get_prev_owner(cls, linked_list: List[str], target: str) -> str:
+        """Given a linked list of strings, it returns the one pointing to target."""
+        root = cls._SENTINEL_OWNERS
+        index = linked_list.index(target) - 1
+        if index < 0:
+            # if target is the first element in the list, the root is pointing to it
+            return root
+        return linked_list[index]
```

Comment:
I would check for the correspondence between the obtained list of owners from the Safe contract and the actual indexing, because it's reverse in the Safe. Please see here:
https://github.com/valory-xyz/autonolas-registries/blob/c3f4e5d9423d7d65e74b248473292e48c437f00c/contracts/multisigs/GnosisSafeSameAddressMultisig.sol#L106-L113

Reply:
Thats a good point. However, I think what we have here is correct. 
https://github.com/safe-global/safe-contracts/blob/v1.3.0/contracts/base/OwnerManager.sol#L136
The order of the owners returned is same as the linked list, ie the first owner in the linked list is the one at index 0.